internal/common/database/types/types.go

```diff
@@ -0,0 +1,118 @@
+
+	// CopyFrom performs a bulk insertion of data into a specified table.
+	// It accepts the table name, column names, and a slice of rows representing the data to be inserted. It returns the number of rows inserted and any error encountered.
+	CopyFrom(ctx context.Context, tableName string, columnNames []string, rows [][]any) (int64, error)
+
+	// Commit commits the transaction. It returns any error encountered during the commit operation.
+	Commit(context.Context) error
+
+	// Rollback rolls back the transaction. It returns any error encountered during the rollback operation.
+	Rollback(context.Context) error
+}
+
+// DatabasePool represents a database connection pool interface that provides methods for acquiring and managing database connections.
+type DatabasePool interface {
+	// Acquire acquires a database connection from the pool.
+	// It takes a context and returns a DatabaseConn representing the acquired connection and any encountered error.
+	Acquire(context.Context) (DatabaseConn, error)
+
+	// Ping pings the database to check the connection. It returns any error encountered during the ping operation.
+	Ping(context.Context) error
+
+	// Close closes the database connection. It returns any error encountered during the closing operation.
+	Close()
+
+	// Exec executes a query that doesn't return rows. It returns any error encountered.
+	Exec(context.Context, string, ...any) (any, error)
+
+	// Query executes a query that returns multiple rows.
+	// It returns a DatabaseRows interface that allows you to iterate over the result set, and any error encountered.
+	Query(context.Context, string, ...any) (DatabaseRows, error)
+
+	// BeginTx starts a transcation with the given DatabaseTxOptions, or returns an error if any occurred.
+	BeginTx(context.Context, DatabaseTxOptions) (DatabaseTx, error)
+
+	// BeginTxFunc starts a transaction and executes the given function within the transaction.
+	// It the function runs successfully, BeginTxFunc commits the transaction, otherwise it rolls back and return an error.
+	BeginTxFunc(context.Context, DatabaseTxOptions, func(DatabaseTx) error) error
+}
+
+// DatabaseRow represents a single row in a result set.
+type DatabaseRow interface {
+	// Scan reads the values from the current row into dest values positionally.
+	// It returns an error if any occurred during the read operation.
+	Scan(dest ...any) error
+}
+
+// DatabaseRows represents an interator over a result set.
+type DatabaseRows interface {
+	// Close closes the result set.
+	Close()
+
+	// Next moves the iterator to the next row in the result set, it returns false if the result set is exhausted, otherwise true.
+	Next() bool
+
+	// Err returns the error, if any, encountered during iteration over the result set.
+	Err() error
+
+	// Scan reads the values from the current row into dest values positionally. It returns an error if any occurred during the read operation.
+	Scan(dest ...any) error
+}
```

Comment:
I think we should add another interface that acts as a generic placeholder for both `DatabasePool` and `DatabaseConn` . In various parts of the armada, such as [here](https://github.com/armadaproject/armada/blob/f041857ec91427da7e133ec0c472c5dd9a4549bd/internal/common/database/migrations.go#L33C5-L33C5) , the `Querier` interface is a placeholder for both `pgxpool.Pool` `pgx.Conn`. It is a loose interface that requires only 3 functions https://pkg.go.dev/github.com/jackc/pgtype@v1.13.0/pgxtype#Querier

You can introduce a new similar one here for `DatabasePool` and `DatabaseConn` that requires maybe only `Exec` and `Query` .